contracts/Marketplace.sol

```diff
@@ -41,19 +43,50 @@ contract Marketplace is Collateral, Proofs {
   }
 
   function myRequests() public view returns (RequestId[] memory) {
-    SetMap.AddressSetMapKey key = _toAddressSetMapKey(msg.sender);
-    return _toRequestIds(activeRequests.values(key));
-  }
-
-  function allRequests() public view returns(RequestId[] memory) {
-    return _toRequestIds(activeRequests.values());
+    SetMap.AddressBytes32SetMapKey key = _toAddressSetMapKey(msg.sender);
+    RequestId[] memory requestIds = _toRequestIds(activeRequestsForClients.values(key));
+    RequestId[] memory result = new RequestId[](requestIds.length);
+    uint8 counter = 0;
+    for (uint8 i = 0; i < requestIds.length; i++) {
+      if (!_isCancelled(requestIds[i])) {
+        result[counter] = requestIds[i];
+        counter++;
+      }
+    }
+    return _toRequestIds(Utils._resize(_toBytes32s(result), counter));
+  }
+
+  function requestsForHost(address host) public view returns(RequestId[] memory) {
+    EnumerableSet.Bytes32Set storage keys = activeRequestsForHosts.keys();
+    uint256 keyLength = keys.length();
+    RequestId[] memory result = new RequestId[](keyLength);
+
+    uint8 counter = 0; // should be big enough
+    for (uint8 i = 0; i < keyLength; i++) {
```

Comment:
Doesn't this iterate through all active requests in the entire network? Even for a view function that seems excessive.
The index `i` is not going to fit in a `uint8` :smile:

Reply:
There is a large trade off between being able to clear all addresses for a request (once a request is cancelled/failed) and being able to view all requests for an address. 

I previously tried using an `AddressBytes32SetMap` which maps `address => RequestIds`. This allows for an efficient view of requests for an address (`requestsForHosts`), however, as mentioned, removing a request once cancelled/failed, requires iteration in a transaction. We could take this a step further by implementing filtering of `RequestIds` based on state inside `requestsForHosts`, which should accurately show requests for a host. However cancelled/failed `RequestIds` would never be removed, thus making the iteration longer and longer as time passes. I was thinking perhaps that we could go even further, doing something similar to `AccountLocks` by removing cancelled/failed requests on each `add`, however, again this requires iteration in a transaction.

Looking at other options...